modules/flowable-engine/src/main/java/org/flowable/engine/impl/delete/ComputeDeleteHistoricProcessInstanceIdsJobHandler.java

```diff
@@ -0,0 +1,593 @@
+                    .listPage(0, 2);
+
+            // We are only going to start deletion if the batch is not failed
+            boolean startDeletion = !DeleteProcessInstanceBatchConstants.STATUS_FAILED.equals(batch.getStatus());
+
+            for (BatchPart nextComputePart : nextComputeParts) {
+                if (!nextComputePart.getId().equals(batchPart.getId())) {
+                    startDeletion = false;
+                    JobService jobService = engineConfiguration.getJobServiceConfiguration().getJobService();
+
+                    JobEntity nextComputeJob = jobService.createJob();
+                    nextComputeJob.setJobHandlerType(ComputeDeleteHistoricProcessInstanceIdsJobHandler.TYPE);
+                    nextComputeJob.setJobHandlerConfiguration(nextComputePart.getId());
+                    jobService.createAsyncJob(nextComputeJob, false);
+                    jobService.scheduleAsyncJob(nextComputeJob);
+                    break;
+                }
+            }
+
+            if (startDeletion) {
+                JobService jobService = engineConfiguration.getJobServiceConfiguration().getJobService();
+                JobEntity nextDeleteJob = jobService.createJob();
+                nextDeleteJob.setJobHandlerType(DeleteHistoricProcessInstanceIdsJobHandler.TYPE);
+                nextDeleteJob.setJobHandlerConfiguration(batchPartForDelete.getId());
+                jobService.createAsyncJob(nextDeleteJob, false);
+                jobService.scheduleAsyncJob(nextDeleteJob);
+            }
+        }
+
+        batchService.completeBatchPart(batchPart.getId(), DeleteProcessInstanceBatchConstants.STATUS_COMPLETED, resultNode.toString());
+    }
+
+    private void failBatchPart(ProcessEngineConfigurationImpl engineConfiguration, BatchService batchService, BatchPart batchPart, Batch batch,
+            String resultJson, boolean sequentialExecution) {
+        batchService.completeBatchPart(batchPart.getId(), DeleteProcessInstanceBatchConstants.STATUS_FAILED, resultJson);
+        if (sequentialExecution) {
+            completeBatch(batch, DeleteProcessInstanceBatchConstants.STATUS_FAILED, engineConfiguration);
+        }
+    }
+
+    protected void completeBatch(Batch batch, String status, ProcessEngineConfigurationImpl engineConfiguration) {
+        engineConfiguration.getBatchServiceConfiguration()
+                .getBatchService()
+                .completeBatch(batch.getId(), status);
+    }
+
+    protected HistoricProcessInstanceQuery createQuery(JsonNode queryNode, ProcessEngineConfigurationImpl engineConfiguration) {
+        HistoricProcessInstanceQuery query = engineConfiguration.getHistoryService()
+                .createHistoricProcessInstanceQuery();
+
+        populateQuery(queryNode, query, engineConfiguration);
+        return query;
+    }
+
+    protected void populateQuery(JsonNode queryNode, HistoricProcessInstanceQuery query, ProcessEngineConfigurationImpl engineConfiguration) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 152 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.